### PR TITLE
add on demand tool 

### PR DESCRIPTION
**Pull request by jerryjliu**

### CODE REVIEW

File: llama_index/tools/adhoc_loader_tool.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            raise ValueError(
+                "Missing query_str in kwargs with parameter name: "
+                f"{self._query_str_kwargs_key}"
+            )
+        if self._use_query_str_in_loader:
+            query_str = kwargs[self._query_str_kwargs_key]
+        else:
+            query_str = kwargs.pop(self._query_str_kwargs_key)
+        docs = self._reader.load_data(*args, **kwargs)
+        index = self._index_cls.from_documents(docs, **self._index_kwargs)
```
Comment (logan-markewich) [inline llama_index/tools/adhoc_loader_tool.py:85]: Part of me says this index should be saved/cached for re-use. But then the other part of me says that doesn't make sense (we don't want to insert the same document more than once from the reader..)

**Review by logan-markewich — APPROVED** [review]
Makes sense to me, and makes the readers a little more automatic. I think I like the term "on demand" more than  "adhoc", not sure which term you wanted to settle on.

Feels like there's room to think about how caching could work in future PRs too!

**Comment by jerryjliu** [comment]
> Makes sense to me, and makes the readers a little more automatic. I think I like the term "on demand" more than "adhoc", not sure which term you wanted to settle on.
> 
> Feels like there's room to think about how caching could work in future PRs too!

makes sense! I think i like on demand too, will revise

File: docs/how_to/integrations/using_with_langchain.md [hunk 2]
```diff
@@ -4,9 +4,18 @@ LlamaIndex provides both Tool abstractions for a Langchain agent as well as a me
...
+### Use any data loader as a Langchain Tool
 
-### Llama Tool abstractions
-LlamaIndex provides Tool abstractions so that you can use LlamaIndex along with a Langchain agent. 
+LlamaIndex allows you to use any data loader within the LlamaIndex core repo or in [LlamaHub](https://llamahub.ai/) as an "on-demand" data query Tool within a LangChain agent.
+
+The Tool will 1) load data using the data loader, 2) index the data, and 3) query the data and return the response in an ad-hoc manner.
+
+**Resources**
+- [OnDemandLoaderTool Tutorial](/examples/tools/OnDemandLoaderTool.ipynb)
```
Comment (Disiok) [inline docs/how_to/integrations/using_with_langchain.md:14]: Cool name. Only other suggestion would be some flavor of `just-in-time`

File: llama_index/tools/function_tool.py [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 from pydantic import BaseModel
 from llama_index.tools.types import BaseTool, ToolMetadata
+from langchain.tools import Tool, StructuredTool
```
Comment (Disiok) [inline llama_index/tools/function_tool.py:5]: a general comment:

would be nice to create a "langchain bridge" module where we route all of all langchain dependencies through. This makes it very easy to have a central place to know all the specific langchain modules we depend on for the future.

File: llama_index/tools/ondemand_loader_tool.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+from llama_index.tools.types import BaseTool, ToolMetadata
+from llama_index.readers.base import BaseReader
+from typing import Any, List, Optional, Dict, Type, Tuple, Callable
+from llama_index.indices.base import BaseGPTIndex
+from llama_index.indices.vector_store import GPTVectorStoreIndex
+from pydantic import BaseModel, create_model
+from inspect import signature
+
+
+def create_schema_from_function(
```
Comment (Disiok) [inline llama_index/tools/ondemand_loader_tool.py:17]: Can we move this to general utility?

File: llama_index/tools/ondemand_loader_tool.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        fields[param_name] = (param_type, param_default)
+
+    additional_fields = additional_fields or []
+    for field_name, field_type, field_default in additional_fields:
+        fields[field_name] = (field_type, field_default)
+
+    return create_model(name, **fields)  # type: ignore
+
+
+class OnDemandLoaderTool(BaseTool):
```
Comment (Disiok) [inline llama_index/tools/ondemand_loader_tool.py:40]: very nit: more like a `OnDemandQueryEngine`? Not sure.

File: llama_index/tools/ondemand_loader_tool.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    """On-demand data loader tool.
+
+    Loads data with query interface, stores in index, and queries
```
Comment (Disiok) [inline llama_index/tools/ondemand_loader_tool.py:43]: very nit: loads data with reader/load?

File: tests/tools/conftest.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
+"""Conftest."""
+
+from typing import List
+import pytest
+
+from llama_index.readers.schema.base import Document
+
+
+@pytest.fixture
+def documents() -> List[Document]:
```
Comment (Disiok) [inline tests/tools/conftest.py:10]: Why are we duplicating this. This already exists under `tests/conftest.py?

File: llama_index/tools/ondemand_loader_tool.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    """
+
+    def __init__(
+        self,
+        reader: BaseReader,
+        index_cls: Type[BaseGPTIndex],
+        index_kwargs: Dict,
+        metadata: ToolMetadata,
+        use_query_str_in_loader: bool = False,
+        query_str_kwargs_key: str = "query_str",
```
Comment (Disiok) [inline llama_index/tools/ondemand_loader_tool.py:55]: these param names not the most intuitive.

File: llama_index/tools/adhoc_loader_tool.py (hunk 1 above)
Comment (Disiok) [reply llama_index/tools/adhoc_loader_tool.py]: Had the same feeling before when we first talked about this. Seems just a bit off that we are loading data and building index just to throw it all away after answer the query.

File: llama_index/tools/function_tool.py (hunk 3 above)
Comment (jerryjliu) [reply llama_index/tools/function_tool.py:5]: yeah that's a good idea...there's also a bunch in `langchain_wrappers`

File: llama_index/tools/ondemand_loader_tool.py (hunk 4 above)
Comment (jerryjliu) [reply llama_index/tools/ondemand_loader_tool.py:17]: yep

File: tests/tools/conftest.py (hunk 7 above)
Comment (jerryjliu) [reply tests/tools/conftest.py:10]: hmm mentioned offline too but i only see it in `tests/indices/conftest.py`

File: llama_index/tools/ondemand_loader_tool.py (hunk 8 above)
Comment (jerryjliu) [reply llama_index/tools/ondemand_loader_tool.py:55]: yeah...will iterate on this probably

**Pull request merged**